File: resources/rekognition.py
import json
import cv2
import os
import boto3
import time
from datetime import datetime


def rekognize(rekog, source_bucket, source_key):
    theCelebs = {}
    strDetail = ""
    strOverall = ""

    startCelebrityrekog = rekog.start_celebrity_recognition(
        Video={
            "S3Object": {
                "Bucket": source_bucket,
                "Name": source_key,
            }
        },
    )

    celebrityJobId = startCelebrityrekog["JobId"]

    print("Rekog Job Id: {0}".format(celebrityJobId))

    # Wait for celebrity recognition job to complete
    # In production use cases, you would usually use StepFucntion or SNS topic to get notified when job is complete.
    getCelebrityRecognition = rekog.get_celebrity_recognition(
        JobId=celebrityJobId, SortBy="TIMESTAMP"
    )

    while getCelebrityRecognition["JobStatus"] == "IN_PROGRESS":
        time.sleep(5)
        print(".", end="")

        getCelebrityRecognition = rekog.get_celebrity_recognition(
            JobId=celebrityJobId, SortBy="TIMESTAMP"
        )

    print(getCelebrityRecognition["JobStatus"])
    print(json.dumps(getCelebrityRecognition, indent=4))

    # Celebrities detected in each frame
    for celebrity in getCelebrityRecognition["Celebrities"]:
        if "Celebrity" in celebrity:
            cconfidence = celebrity["Celebrity"]["Confidence"]
            if cconfidence > 95:
                ts = celebrity["Timestamp"]
                cname = celebrity["Celebrity"]["Name"]
                strDetail = strDetail + "At {} ms: {} (Confidence: {})\n".format(
                    ts, cname, round(cconfidence, 2)
                )
                if not cname in theCelebs:
                    theCelebs[cname] = cname

    # Unique faces detected in video
    for theCeleb in theCelebs:
        strOverall = strOverall + "Name: {}\n".format(theCeleb)

    print(strDetail)
    print(strOverall)

    return theCelebs


def handler(event, context):
    s3 = boto3.client("s3")
    rekog = boto3.client("rekognition")
    dynamodb = boto3.resource("dynamodb")

    # Get the dynamodb table
    dynamodb_name = os.environ.get("TABLE")
    table = dynamodb.Table(dynamodb_name)
    print("[JINLOG] Successfully load the table " + dynamodb_name)

    # Get the bucket and key from the MediaConvert event
    file_info = event["detail"]["outputGroupDetails"][1]["outputDetails"][0][
        "outputFilePaths"
    ][0].split("//")
    file_info = file_info[1].split("/")

    bucket_name = file_info[0]
    object_key = ""
    for i in range(1, len(file_info) - 1):
        object_key = file_info[i] + "/"
    object_key += file_info[i + 1]

    # Call Rekognition for object detections
    celebrities = rekognize(rekog, bucket_name, object_key)

    if 'Son Heung-min' in celebrities : 

        playlist_filepath = event["detail"]["outputGroupDetails"][0]["playlistFilePaths"][0]

        # Put an item in the table
        table.put_item(
            Item={
                "filepath": playlist_filepath,
                "datetime": datetime.now().strftime("%Y-%m-%d-%H:%M:%S"),
                "celebrities": celebrities,
            }
        )

File: resources/rekognition.ts
import { DynamoDBClient, PutItemCommand } from "@aws-sdk/client-dynamodb";
import {
  RekognitionClient,
  StartCelebrityRecognitionCommand,
  GetCelebrityRecognitionCommand,
} from "@aws-sdk/client-rekognition";
import { EventBridgeEvent, S3Event } from "aws-lambda";

interface ISvgObj {
  [key: string]: string;
}

export const handler = async (event: EventBridgeEvent<"Media Convertin Complete Event", any>) => {
  const rekognition = new RekognitionClient({ region: "ap-northeast-2" });
  const dynamodb = new DynamoDBClient({ region: "ap-northeast-2" });

  // Get dynamoDB table name from environment variables
  const tableName = process.env["TABLE"];

  console.log(`Event received from EventBridge: ${JSON.stringify(event)}`);

  // Get the bucket and key name from the event
  let objectInfo = event["detail"]["outputGroupDetails"][1]["outputDetails"][0]["outputFilePaths"][0].split("//");
  objectInfo = objectInfo[1].split("/");

  const bucket = objectInfo[0];
  let key = objectInfo[1];
  let i = 2;
  while (i < objectInfo.length) {
    key += "/" + objectInfo[i];
    i++;
  }

  console.log(`Bucket: ${bucket}, Key: ${key}`);

  const celebrities = await rekognize(rekognition, bucket, key);

  if ("Son Heung-min" in celebrities) {
    const videoPath = "s3://" + bucket + "/" + key;

    dynamodb.send(
      new PutItemCommand({
        TableName: tableName,
        Item: {
          PK: { S: videoPath },
          SK: { S: Date.now().toString() },
          celebs: { S: JSON.stringify(celebrities) },
        },
      })
    );
  }
};

const rekognize = async (rekog: RekognitionClient, bucket: string, key: string) => {
  let theCelebs: ISvgObj = {};

  const startRekogCelebRes = await rekog.send(
    new StartCelebrityRecognitionCommand({
      Video: {
        S3Object: {
          Bucket: bucket,
          Name: key,
        },
      },
    })
  );

  // Get Job Id
  const jobId = startRekogCelebRes["JobId"];
  console.log(`Rekognition Job Id: ${jobId}`);

  let getRekogCelebRes = await rekog.send(
    new GetCelebrityRecognitionCommand({
      JobId: jobId,
      SortBy: "TIMESTAMP",
    })
  );

  // if Rekognition job is still in progress, sleep for 5 seconds
  // and try it again
  while (getRekogCelebRes.JobStatus == "IN_PROGRESS") {
    console.log("Rekognition job in progress...", getRekogCelebRes.Celebrities);
    await new Promise((f) => setTimeout(f, 5000));
    getRekogCelebRes = await rekog.send(
      new GetCelebrityRecognitionCommand({
        JobId: jobId,
        SortBy: "TIMESTAMP",
      })
    );
  }

  console.log(getRekogCelebRes.JobStatus);
  console.log(JSON.stringify(getRekogCelebRes));

  for (const celebrity of getRekogCelebRes.Celebrities!) {
    let strDetail = "";
    if (celebrity.Celebrity !== undefined) {
      const cconfidence = celebrity["Celebrity"]["Confidence"];
      if (cconfidence !== undefined && cconfidence > 95) {
        const ts = celebrity.Timestamp;
        const name = celebrity.Celebrity.Name!;
        strDetail += strDetail + `At ${ts} ms: ${name} (confidence ${Math.round(cconfidence)})`;
        if (!(name in theCelebs)) {
          theCelebs[name] = name;
        }
      }
    }
  }
  return theCelebs;
};
